feat:  Allow review rejections to be undone

pselkirk opened the pull request:
Addresses most of #5978 and part of #6169

pselkirk committed:
feat: Allow reviewer to accept a review they're previously rejected

rjsparks commented:
There's at least one rough edge (and this all needs tests).

I ran through having cellar-flac assigned to me, and then rejecting/accepting/rejecting a few times. In my dev instance I've rejected the assignment there are no other assignements, and the open assignments management page doesn't allow the secretary to assign it to me again. The request, as I understand it, is to have me show up on the list here, but marked "has previously rejected reviewing this doc":

![image](https://github.com/ietf-tools/datatracker/assets/10996692/59daabee-a64e-457c-a3db-e151cc50b3c8)

(right side truncated to not overshare current states)

rjsparks commented:
@russhousley @ajeanmahoney @kivinen 

This is what the assignment page will look like for a reviewer that has rejected a review before. This is Russ' view after both of us have been assigned and rejected the document.

Is the UI here intuitive enough? And if you have this as secretaries (if you were logged in, Jean, you would see accept buttons for both of us), would you be ok _not_ having the person show up in the list above?

![image](https://github.com/ietf-tools/datatracker/assets/10996692/7294d1d3-34df-423a-a98c-aa4d0213eda1)

pselkirk committed:
feat: Add a reviewer who has previously rejected a review to the list of suggested reviewers.

This largely un-does d105f8b, at the request of at least one team secretary.

pselkirk committed:
fix: Went a little overboard on the previous commit

one_assignment still has to exclude reviewers who rejected the assignment,
or they could end up being the suggested reviewer.

pselkirk committed:
fix: Actually do the assignment

pselkirk committed:
fix: If there's an existing assignment, don't create a new one

ajeanmahoney commented:
I would typically work from the list on the **Manage unassigned open review requests** page rather than an individual review request page. As secretary, I would expect to see all available reviewers. 

On the individual review request page, the secretary is reassigning to Russ, yes?

Maybe add a date to "Assignment rejected" under the reviewer name, e.g., "Assignment previously rejected 18 Sept 2023". Maybe there should be a **Reject again** button?

rjsparks commented on ietf/review/policies.py:
```diff
@@ -70,6 +69,12 @@ def assign_reviewer(self, review_req, reviewer, add_skip):
+            assignment.assigned_on = timezone.now()
+            assignment.save()
+            return assignment
         return review_req.reviewassignment_set.create(state_id='assigned', reviewer=reviewer, assigned_on=timezone.now())
```
Consider putting this in an `else:` clause

pselkirk committed:
style: Restructure conditional for clarity

pselkirk committed:
test: Add test cases for accepting or assigning a review assignment after rejecting it
